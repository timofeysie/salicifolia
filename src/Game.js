import React from 'react'; // import React, { Component } from 'react';
import Board from './Board.js';

export class Game extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            history: [{
                squares: Array(9).fill(null),
            }],
            xIsNext: true,
        };
    }

    handleClick(i) {
        const history = this.state.history;
        const current = history[history.length - 1];
        // call .slice() to copy the squares array
        // instead of mutating 
        const squares = current.squares.slice();
        if (this.calculateWinner(squares) || squares[i]) {
            return;
        }
        squares[i] = this.state.xIsNext ? 'X' : 'O';
        this.setState({
            history: history.concat([{
              squares: squares,
            }]),
            xIsNext: !this.state.xIsNext,
        });
    }

    calculateWinner(squares) {
        const lines = [
          [0, 1, 2],
          [3, 4, 5],
          [6, 7, 8],
          [0, 3, 6],
          [1, 4, 7],
          [2, 5, 8],
          [0, 4, 8],
          [2, 4, 6],
        ];
        for (let i = 0; i < lines.length; i++) {
          const [a, b, c] = lines[i];
          if (squares[a] && squares[a] === squares[b] && squares[a] === squares[c]) {
            return squares[a];
          }
        }
        return null;
      }
    
    render() {
        const history = this.state.history;
        const current = history[history.length - 1];
        const winner = this.calculateWinner(current.squares);
        let status;
        if (winner) {
            status = 'Winner: ' + winner;
        } else {
            status = 'Next player: ' + (this.state.xIsNext ? 'X' : 'O');
        }
        return (
            <div className="game">
                <div className="game-board">
               <Board 
                    squares={current.squares}
                    onClick={(i) => this.handleClick(i)}
                />
                </div>
            <div className="game-info">
                <div>{status}</div>
                <ol>{/* TODO */}</ol>
            </div>
            </div>
      );
    }
}

export default Game;
  