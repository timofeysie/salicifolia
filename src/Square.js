import React from 'react';
/**
 * Controlled components
 * It receives its value from its parent Board 
 * and informs its parent when itâ€™s clicked.
 */
class Square extends React.Component {
    render() {
      return (
        <button className="square" 
            onClick={() => {
                this.props.onClick();
            }}>
          {this.props.value}
        </button>
      );
    }
}

export default Square;
