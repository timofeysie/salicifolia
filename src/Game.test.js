import React from 'react';
import Game from './Game';
import Board from './Board.js';
import Square from './Square.js';
import renderer from 'react-test-renderer';
import Enzyme from 'enzyme';
import {shallow} from 'enzyme';
import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
configure({ adapter: new Adapter() });

test('tree to Match Snapshot', () => {
  const component = renderer.create(
    <Game />,
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();

  // manually trigger the callback
  // tree.props.handleClick(0);
  // // re-rendering
  // tree = component.toJSON();
  // expect(tree).toMatchSnapshot();
  //console.log(tree.children);

});

test('Game board is created', () => {

  // Render a checkbox with label in the document
  const game = shallow(<Game />);
  console.log('game',game);
  expect(game.find(Board)).to.have.length(1);
  expect(game.find(Square)).to.have.length(6);

  
  //expect(checkbox.text()).toEqual('Off');
  //checkbox.find('input').simulate('change');
});