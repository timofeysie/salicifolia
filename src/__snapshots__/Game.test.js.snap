// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tree to Match Snapshot 1`] = `
<div
  className="game"
>
  <div
    className="game-board"
  >
    <div>
      <div
        className="board-row"
      >
        <button
          className="square"
          onClick={[Function]}
        />
        <button
          className="square"
          onClick={[Function]}
        />
        <button
          className="square"
          onClick={[Function]}
        />
      </div>
      <div
        className="board-row"
      >
        <button
          className="square"
          onClick={[Function]}
        />
        <button
          className="square"
          onClick={[Function]}
        />
        <button
          className="square"
          onClick={[Function]}
        />
      </div>
      <div
        className="board-row"
      >
        <button
          className="square"
          onClick={[Function]}
        />
        <button
          className="square"
          onClick={[Function]}
        />
        <button
          className="square"
          onClick={[Function]}
        />
      </div>
    </div>
  </div>
  <div
    className="game-info"
  >
    <div>
      Next player: X
    </div>
    <ol>
      <li>
        <button
          onClick={[Function]}
        >
          Go to game start
        </button>
      </li>
    </ol>
  </div>
</div>
`;
